test/unit/puppetlabs/puppetserver/ruby/http_client_test.clj

```diff
@@ -142,98 +151,144 @@
 
 (deftest connections-closed
   (testing "connection header always set to close on get"
     (logutils/with-test-logging
       (jetty9/with-test-webserver ring-app-connection-closed port
-        (let [scripting-container (create-scripting-container port)]
-          (is (= "The Connection header has value close" (.runScriptlet scripting-container "c.get('/', {}).body")))))))
-  (testing "connection header always set to close on get"
+        (let [sc (create-scripting-container port)]
+          (is (= "The Connection header has value close" (.runScriptlet sc "c.get('/', {}).body")))
+          (terminate-scripting-container sc)))))
+  (testing "connection header always set to close on post"
     (logutils/with-test-logging
       (jetty9/with-test-webserver ring-app-connection-closed port
-        (let [scripting-container (create-scripting-container port)]
-          (is (= "The Connection header has value close" (.runScriptlet scripting-container "c.post('/', 'foo', {}).body"))))))))
+        (let [sc (create-scripting-container port)]
+          (is (= "The Connection header has value close" (.runScriptlet sc "c.post('/', 'foo', {}).body")))
+          (terminate-scripting-container sc)))))
+  (testing "client's terminate function closes the client"
+    (logutils/with-test-logging
+      (jetty9/with-test-webserver ring-app-connection-closed port
+        (let [sc (create-scripting-container port)]
+          (.runScriptlet sc "response = c.get('/', {})")
+          (is (= "200" (.runScriptlet sc "response.code")))
+          (.runScriptlet sc "c.class.terminate")
+          (try
+            (.runScriptlet sc "response = c.get('/', {})")
+            (catch EvalFailedException e
+              (let [wrapped-exception (.getCause e)
+                    message           (.getMessage e)]
+                (is (instance? IllegalStateException wrapped-exception))
+                (is (re-find #"Request cannot be executed; I/O reactor status: STOPPED" message)))))
+          (.terminate sc))))))
+
+
+(deftest http-and-https
+  (testing "can make http and https requests without having to use a new scripting container"
+    (logutils/with-test-logging
+      (jetty9/with-test-webserver ring-app-alternate port
+        (with-webserver-with-protocols nil nil
+          (let [sc (create-scripting-container 10080 {:use-ssl true})]
+            (.runScriptlet sc "response = c.get('/', {})")
+            (is (= "200" (.runScriptlet sc "response.code")))
+            (is (= "hi" (.runScriptlet sc "response.body")))
+            (.runScriptlet sc (str (format "c = Puppet::Server::HttpClient.new('localhost', %d, {:use_ssl => %s});"
+                                           port false)
+                                   "response = c.get('/', {})"))
+            (is (= "200" (.runScriptlet sc "response.code")))
+            (is (= "bye" (.runScriptlet sc "response.body")))
+            (terminate-scripting-container sc)))))))
```

Comment:
I think this is a good test although what led to the creation of this test was the concern that if an HTTP request is made before the HTTPS one that the HTTPS one would fail because the SSLContext wasn't configured when the client was first created.  I think the currently written test covers HTTPS before HTTP.  Might be good to add a second HTTP before HTTPS test to ensure both cases are covered.